docs/lang/control-and-nondeterminism.md

```diff
@@ -0,0 +1,509 @@
+we can think of `GUESS S` as being one of the following implementations:
+
+ 1. `GUESS S` can be a remote procedure call in a distributed system.  Unless,
+ we have centralized control over the distributed system, the returned value of
+ RPC may be non-deterministic.
+
+ 1. `GUESS S` can be simply the user input. In this case, the user resolves
+ non-determinism.
+
+ 1. `GUESS S` can be controlled by an adversary, who is trying to break the
+ system.
+
+ 1. `GUESS S` can pick an element by calling a random number generator.
+ However, note that RNG is a very special way of resolving non-determinism: It
+ assumes probabilistic distribution of elements (usually, it is close to the
+ [uniform
+ distribution](https://en.wikipedia.org/wiki/Discrete_uniform_distribution)).
+ Thus, the probability of producing an unfair choice of elements with RNG will
+ be approaching 0.
+
+
+As you see, there are multiple sources of non-determinism. With `GUESS S` we can
+model all of them. As TLA+ does not introduce special primitives for different
+kinds of non-determinism, neither do we fix any implementation of `GUESS S`.
+
+**Halting.** Note that `GUESS {}` halts the evaluation. What does it mean? The
+evaluation cannot continue in the current partial state. It does not imply that
+we have found a deadlock in our TLA+ specification. It simply means that we
+made wrong choices on the way. If we like to enumerate all possible state
+successors, like TLC does, we have to backtrack. In general, the course of
+action depends on the program analysis that you implement. For instance,
+a random simulator could simply backtrack and randomly choose another value.
+
+<a name="nondetExists"></a>
+### Non-determinism in `\E x \in S: P`
+
+We only have to consider the following case: `\E x \in S: P` is evaluated in a
+partial state `s`, and there is a primed state variable `y'` that satisfies two
+conditions:
+
+ 1. The predicate `P` refers to `y'`, that is, `P` has to assign a value to `y'`.
+ 2. The value of `y'` is not defined yet, that is, `s.y' = Null`.
+
+If the above assumptions do not hold true, the expression `\E x \in S: P` does
+not have non-determinism and it can be evaluated by following the standard
+deterministic semantics of exists, see [[Logic]](./logic.md).
+
+Now it is very easy to evaluate `\E x \in S: P`. We simply evaluate the
+following expression:
+
+```tla
+  LET x == GUESS S IN P
+```
+
+It is the job of `GUESS S` to tell us what value of `x` should be
+evaluated. There are three possible outcomes:
+
+ 1. Predicate `P` evaluates to `TRUE` when using the provided value of `x`.
+ In this case, `P` assigns the value of an expression `e` to `y'` as soon as
+ the evaluator meets the expression `y' = e`.
+ The evaluation may continue.
+ 1. Predicate `P` evaluates to `FALSE` when using the provided value of `x`.
+ Well, that was a wrong guess. According to our semantics, the evaluation
+ halts. See the above discussion on "halting".
```

Comment:
The above discussion does not go into depth on this. What happens is that the next guess is made as follows:
`GUESS S \{x} IN P ` until `S` is exhausted. Only then is the entire `\E` expression evaluated as `FALSE`.

Reply:
I have a simpler semantics in mind. It is not about enumerating all states but like a non-deterministic Turing machine.